File: Backend/Nodejs/src/controllers/song.ts
import { Request, Response } from "express";
import { FavoriteSong, Song, SongFiles, UpdateSong } from "../models/types";
import { SongModel } from "../models/song";
import { checkKeys } from "../utils/checkKeys";

export class SongController {
  static createSong(req: Request, res: Response) {
    const song = req.body as Song;
    const files = req.files as SongFiles;

    if (
      !checkKeys(song, ["cover", "source"]) || !("cover" in files) ||
      !("source" in files)
    ) {
      return res.status(400).json({ MESSAGE: "Faltan parámetros" });
    }
    SongModel.createSong(song, files, (response: any, ok: Boolean) => {
      // Respuesta
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static getSong(req: Request, res: Response) {
    const { id } = req.params;
    if (!id) return res.status(401).json({ message: "Falta el id" });
    const id_song = parseInt(id);
    SongModel.getSong({ id: id_song }, (response: any, ok: Boolean) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static getSongs(req: Request, res: Response) {
    SongModel.getSongs((response: any, ok: Boolean) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static editSong(req: Request, res: Response) {
    const song = req.body as UpdateSong;
    const files = req.files as SongFiles;
    
    SongModel.editSong(song, files, (response: string, ok: Boolean) => {
      // Respuesta
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static deleteSong(req: Request, res: Response) {
    const { id } = req.params;
    if (!id) return res.status(401).json({ MESSAGE: "Falta el id" });
    const id_album = parseInt(id);
    SongModel.deleteSong({ id: id_album }, (response, ok) => {
      res.status(ok ? 200 : 400).json(response);
    });
  }

  static addToFavorites(req: Request, res: Response) {
    const song = req.body as FavoriteSong;
    // Validar datos
    if (!checkKeys(song)) {
      return res.status(400).json({ message: "Faltan datos" });
    }
    SongModel.addToFavorite(song, (response: any, ok: Boolean) => {
      // Respuesta
      res.status(ok ? 200 : 400).json(response);
    });
  }
}

File: Backend/Python/src/controllers/song.py
from flask import request
from src.libs.s3client import upload_file
from src.models.song import SongModel


class SongController:

    @staticmethod
    def create_song():
        body = request.form
        file = request.files
        if len(body) == 0 or len(file) == 0:
            return {"MESSAGE": "Faltan datos", "TYPE": "ERROR"}, 401
        cover_location = upload_file("song", file["cover"])
        if cover_location == "":
            return {"MESSAGE": "No se pudo subir la foto", "TYPE": "ERROR"}, 401
        sourcer_location = upload_file("song", file["source"], False)
        if sourcer_location == "":
            return {"MESSAGE": "No se pudo subir la canción", "TYPE": "ERROR"}, 401
        response = SongModel.create_song(body, cover_location, sourcer_location)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def get_song(id, email):
        response = SongModel.get_song(id, email)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def get_songs():
        response = SongModel.get_songs()
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def delete_song(id):
        response = SongModel.delete_song(id)
        return response[0], (200 if response[1] else 400)

    @staticmethod
    def edit_song():
        song = dict(request.form)
        files = request.files
        if len(song) == 0 or len(files) == 0:
            return {"MESSAGE": "Faltan datos"}, 400
        cover_location = upload_file("song", files["cover"])
        sourcer_location = upload_file("song", files["source"], False)
        # Guardar en db
        response = SongModel.edit_song(song, cover_location, sourcer_location)
        return response[0], (200 if response[1] else 400)
